[TySan] User-friendly (C style) pointer type names for error reports

BStott6 opened the pull request:
Changed the TypeSanitizer instrumentation pass to parse TBAA's `pN T` (e.g. `p2 int`) pointer type names and rewrite them in a more user-familiar `T*` notation. Updated TySan docs to remove the explanation for the strange pointer type names. Updated TySan regression tests which refer to the pointer type formatting to match the new formatting. Added a command line option in `TypeSanitizer.cpp` to use the old TBAA-style type names instead (`tysan-use-tbaa-type-names`).

BStott6 committed:
[TySan] User-friendly (C style) pointer type names for error reports

gbMattN commented on clang/docs/TypeSanitizer.rst:
```diff
@@ -119,8 +119,6 @@ brief dictionary of these terms.
 
 * ``omnipotent char``: This is a special type which can alias with anything. Its name comes from the C/C++ 
   type ``char``.
-* ``type p[x]``: This signifies pointers to the type. ``x`` is the number of indirections to reach the final value.
```
If we are always doing C style maybe it would make sense then to change "omnipotent char" to char as well? That or change the docs to state the name is permanent

BStott6 committed:
Fix failing test

BStott6 committed:
Remove command line flag, fix formatting

BStott6 committed:
Rework pointer typename rewriting to occur in runtime rather than instrumentation, fix demangling for pointer names

BStott6 committed:
Fix test

BStott6 committed:
Add regression test for pointer typename formatting

BStott6 committed:
Refactor type name demangling to write into buffer

BStott6 committed:
Add comment explaining purpose of writeDemangledTypeName

BStott6 commented:
In order to be able to concatenate the pointer asterisks onto the type name without allocating, I have changed `getDisplayName` to `writeDemangledTypeName` which writes the name into a provided buffer. Originally I had changed it to write the name straight to stdout, but this would be incompatible with @gbMattN's PR https://github.com/llvm/llvm-project/pull/169023.

BStott6 commented on compiler-rt/lib/tysan/tysan.cpp:
```diff
@@ -40,20 +41,62 @@ tysan_copy_types(const void *daddr, const void *saddr, uptr size) {
+    return {0, Name};
+
+  return {Indirection, Name + CharIndex};
+}
```
> Can we use some existing utilities (perhaps `std::from_chars` or from `StringRef`) to avoid manually parsing numbers?

As I understand it, use of standard library functions in the sanitizers is not allowed as users may not be linking to the (C or C++) standard libraries, and interceptors may erroneously intercept any calls. However, there is an existing function `internal_simple_strtoll` in `sanitizer_libc` which can be used for this purpose 👍

BStott6 committed:
Refactor parseIndirectionPrefix to use internal_simple_strtoll

BStott6 committed:
Drop braces

Co-authored-by: Florian Hahn <flo@fhahn.com>

BStott6 committed:
Replace stack allocated buffer with vector to eliminate length limit, add test case for double indirection

fhahn reviewed (approved):
Overall LGM, but not sure about the common practices for sharing helpers between sanitizers, so another review would be good .

gbMattN committed:
Rearange vsnprintf

gbMattN committed:
Merge branch 'main' into tysan-better-pointer-formatting

fmayer commented on compiler-rt/lib/sanitizer_common/sanitizer_common.h:
```diff
@@ -15,6 +15,8 @@
 #ifndef SANITIZER_COMMON_H
 #define SANITIZER_COMMON_H
 
+#include <stdarg.h>
```
I don't think we should be including libc here

gbMattN commented on compiler-rt/lib/sanitizer_common/sanitizer_printf.h:
```diff
@@ -0,0 +1,24 @@
+//===-- sanitizer_printf.h ------------------------------------------------===//
```
@fmayer it was here due to including stdarg, mentioned in comments below. Is there a good place to put it if it requires that library?

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_vector.h:
Should be a separate PR?

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_printf.cpp:
should be a separate PR
